azure-iot-device/tests/common/pipeline/pipeline_stage_test.py

```diff
@@ -14,29 +14,202 @@
+):
+    class StageTestConfig(stage_test_config_class):
+        @pytest.fixture
+        def stage(self, mocker, cls_type, init_kwargs):
+            stage = cls_type(**init_kwargs)
+            stage.next = mocker.MagicMock()
+            stage.previous = mocker.MagicMock()
+            mocker.spy(stage, "send_op_down")
+            mocker.spy(stage, "send_event_up")
+            return stage
+
+    #######################
+    # INSTANTIATION TESTS #
+    #######################
+
+    @pytest.mark.describe("{} -- Instantiation".format(stage_class_under_test.__name__))
+    class StageBaseInstantiationTests(StageTestConfig):
+        @pytest.mark.it("Initializes 'name' attribute as the classname")
+        def test_name(self, cls_type, init_kwargs):
+            stage = cls_type(**init_kwargs)
+            assert stage.name == stage.__class__.__name__
+
+        @pytest.mark.it("Initializes 'next' attribute as None")
+        def test_next(self, cls_type, init_kwargs):
+            stage = cls_type(**init_kwargs)
+            assert stage.next is None
+
+        @pytest.mark.it("Initializes 'previous' attribute as None")
+        def test_previous(self, cls_type, init_kwargs):
+            stage = cls_type(**init_kwargs)
+            assert stage.previous is None
+
+        @pytest.mark.it("Initializes 'pipeline_root' attribute as None")
+        def test_pipeline_root(self, cls_type, init_kwargs):
+            stage = cls_type(**init_kwargs)
+            assert stage.pipeline_root is None
+
+    if extended_stage_instantiation_test_class:
+
+        class StageInstantiationTests(
+            extended_stage_instantiation_test_class, StageBaseInstantiationTests
+        ):
+            pass
+
+    else:
+
+        class StageInstantiationTests(StageBaseInstantiationTests):
+            pass
+
+    setattr(
+        test_module,
+        "Test{}Instantiation".format(stage_class_under_test.__name__),
+        StageInstantiationTests,
+    )
+
+    ##############
+    # FLOW TESTS #
+    ##############
+
+    @pytest.mark.describe("{} - .send_op_down()".format(stage_class_under_test.__name__))
+    class StageSendOpDownTests(StageTestConfig):
+        @pytest.mark.it("Completes the op with failure (PipelineError) if there is no next stage")
+        def test_fails_op_when_no_next_stage(self, mocker, stage, arbitrary_op):
+            stage.next = None
+
+            assert not arbitrary_op.completed
+
+            stage.send_op_down(arbitrary_op)
+
+            assert arbitrary_op.completed
+            assert type(arbitrary_op.error) is pipeline_exceptions.PipelineError
+
+        @pytest.mark.it("Passes the op to the next stage's .run_op() method")
+        def test_passes_op_to_next_stage(self, mocker, stage, arbitrary_op):
+            stage.send_op_down(arbitrary_op)
+            assert stage.next.run_op.call_count == 1
+            assert stage.next.run_op.call_args == mocker.call(arbitrary_op)
```

Comment:
>assert stage.next.run_op.call_args == mocker.call(arbitrary_op) [](start = 12, length = 63)

observation: we should use this style (assert call_args == mocker_call(blah)) as often as possible.  Checking all args is much better than checking individual args.  (unless we have to like in test_no_previous_stage) below.  mocker.ANY can be used in cases where we don't want to assert a specific parameter value.

Reply:
I fully agree, this is something I always try to do. It's more accurate, and more readable.

---
In reply to: [354875237](https://github.com/Azure/azure-iot-sdk-python/pull/409#discussion_r354875237) [](ancestors = 354875237)